protobuf-format.md

```diff
@@ -0,0 +1,227 @@
+of that specification.
+
+In this document the terms *Protocol Buffers*, *protobuf*, and *proto* are used
+interchangeably.
+
+## Status of this document
+
+This document is a working draft.
+
+## Table of Contents
+
+1. [Introduction](#1-introduction)
+2. [Attributes](#2-attributes)
+3. [Data](#3-data)
+4. [Transport](#4-transport)
+5. [Examples](#5-examples)
+
+## 1. Introduction
+
+[CloudEvents][ce] is a standardized and protocol-agnostic definition of the
+structure and metadata description of events. This specification defines how the
+elements defined in the CloudEvents specification are are represented using
+a protobuf schema.
+
+The [Attributes](#2-attributes) section describes the naming conventions and
+data type mappings for CloudEvents attributes for use as protobuf message
+properties.
+
+The [Data](#3-data) section describes how the event payload is carried.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][rfc2119].
+
+### 1.2 Content-Type
+
+There is no official IANA *media-type* designation for protobuf, as such this
+specification uses 'application/protobuf' to identify such content.
+
+## 2. Attributes
+
+This section defines how CloudEvents attributes are represented in the protobuf
+[schema][proto-schema].
+
+## 2.1 Type System
+
+The CloudEvents type system is mapped to protobuf as follows :
+
+| CloudEvents   | protobuf |
+| ------------- | ---------------------------------------------------------------------- |
+| Boolean       | [boolean][proto-scalars] |
+| Integer       | [sint32][proto-scalars] |
+| String        | [string][proto-scalars] |
+| Binary        | [bytes][proto-scalars] |
+| URI           | [string][proto-scalars] following [RFC 3986 §4.3][rfc3986-section43]|
+| URI-reference | [string][proto-scalars] following [RFC 3986 §4.1][rfc3986-section41] |
+| Timestamp     | [Timestamp][proto-timestamp]  |
+
+## 2.3 Required Attributes
+
+Required attributes are represented explicitly as protobuf fields.
+
+## 2.4 Optional Attributes & Extensions
+
+Optional and extension attributes are represented using a map construct enabling
+direct support of the CloudEvent [type system][ce-types].
+
+```proto
+map<string, CloudEventAttribute> attribute = 1;
+
+message CloudEventAttribute {
+
+    oneof attr_oneof {
+      bool ce_boolean = 1;
+      sfixed32 ce_integer = 2;
+      string ce_string = 3;
+      bytes ce_binary = 4;
+      string ce_uri = 5;
+      string ce_uri_reference = 6;
+      google.protobuf.Timestamp ce_timestamp = 7;
+    }
+}
+```
+
+In this model an attribute's name is used as the map *key* and is
+associated with its *value* stored in the appropriately typed property.
+
+This approach allows for attributes to be represented and transported
+with no loss of *type* information.
```

Comment:
Unlike the JSON format, presumably - where for an extension attribute (where we don't necessarily know the type inherently in the way that we do for the built-in attributes) we don't know whether a given attribute value is a string that also happens to be a URI but shouldn't be treated as one, or is *really* intended to be a URI. (Hope that makes sense...)

I think we're missing some detail in terms of requirements of a CloudEvent format - when designing one, it would be nice to have guidance about what's required, desirable etc.